File: node/storage-cleaner/src/appwrite.js
import { Client, Storage, Query } from 'node-appwrite';
import { getExpiryDate } from './util.js';

class AppwriteService {
  constructor(apiKey) {
    const client = new Client()
      .setEndpoint(process.env.APPWRITE_FUNCTION_API_ENDPOINT)
      .setProject(process.env.APPWRITE_FUNCTION_PROJECT_ID)
      .setKey(apiKey);
    this.storage = new Storage(client);
  }

  /**
   * Clean up files from the storage bucket by removing files older than a specified retention period.
   *
   * @param {string} bucketId - The ID of the storage bucket to clean.
   * @returns {Promise<void>} A Promise that resolves when the bucket is cleaned.
   */
  async cleanBucket(bucketId) {
    let response;
    const queries = [
      Query.lessThan('$createdAt', getExpiryDate()),
      Query.limit(25),
    ];
    do {
      response = await this.storage.listFiles(bucketId, queries);
      await Promise.all(
        response.files.map((file) =>
          this.storage.deleteFile(bucketId, file.$id)
        )
      );
    } while (response.files.length > 0);
  }
}

export default AppwriteService;

File: python/storage-cleaner/src/appwrite.py
"""
Appwrite service module to handle storage cleanup operations.
"""

import os
from appwrite.client import Client
from appwrite.services.storage import Storage
from appwrite.query import Query
from .utils import get_expiry_date


class AppwriteService:
    """
    Service class to interact with Appwrite's storage service.
    """

    def __init__(self, api_key: str):
        client = (
            Client()
            .set_endpoint(os.getenv("APPWRITE_FUNCTION_API_ENDPOINT"))
            .set_project(os.getenv("APPWRITE_FUNCTION_PROJECT_ID"))
            .set_key(api_key)
        )
        self.storage = Storage(client)

    def clean_bucket(self, bucket_id: str):
        """
        Clean up files from the storage bucket by removing files older than a
        specified retention period.

        :param bucket_id: The ID of the storage bucket to clean.
        """
        queries = [
            Query.less_than("$createdAt", get_expiry_date()),
            Query.limit(25),
        ]

        deleted_files_count = 0
        failed_files = []

        while True:
            try:
                response = self.storage.list_files(bucket_id, queries)
            except Exception as e:
                raise RuntimeError(
                    f"Failed to list files from bucket {bucket_id}: {str(e)}"
                ) from e
            files = response.get("files", [])

            for f in files:
                try:
                    file_id = f.get("$id")
                    if file_id:
                        self.storage.delete_file(bucket_id, file_id)
                        deleted_files_count += 1
                except Exception as e:
                    failed_files.append({"id": file_id, "error": str(e)})

            if not files:
                break

        if failed_files:
            raise RuntimeError(
                f"Deleted {deleted_files_count} files, but failed to delete {len(failed_files)} files: {failed_files}"
            )
